pkg/cache/sql/informer/listoption_indexer.go

```diff
@@ -427,38 +534,179 @@ func (l *ListOptionIndexer) validateColumn(column string) error {
 
-		orWhereClause += fmt.Sprintf(`f."%s" %s ? ESCAPE '\'`, columnName, opString)
-		format := strictMatchFmt
+//type Filter struct {
+//	Field   []string
+//	Match   []string
+//	Op      Op: one of
+//	Partial bool
+//}
+
+// Possible ops from the k8s parser:
+// KEY = and == (same) VALUE
+// KEY != VALUE
+// KEY exists []  # ,KEY, => this filter
+// KEY ! []  # ,!KEY, => assert KEY doesn't exist
+// KEY in VALUES
+// KEY notin VALUES
+
+func (l *ListOptionIndexer) getFieldFilter(filter Filter) (string, []any, error) {
+	opString := ""
+	escapeString := ""
+	columnName := toColumnName(filter.Field)
+	if err := l.validateColumn(columnName); err != nil {
+		return "", nil, err
+	}
+	switch filter.Op {
+	case Eq:
 		if filter.Partial {
-			format = matchFmt
-		}
-		match := filter.Match
-		// To allow matches on the backslash itself, the character needs to be replaced first.
-		// Otherwise, it will undo the following replacements.
-		match = strings.ReplaceAll(match, `\`, `\\`)
-		match = strings.ReplaceAll(match, `_`, `\_`)
-		match = strings.ReplaceAll(match, `%`, `\%`)
-		params = append(params, fmt.Sprintf(format, match))
-		if index == len(orFilters.Filters)-1 {
-			continue
+			opString = "LIKE"
+			escapeString = escapeBackslashDirective
+		} else {
+			opString = "="
+		}
+		clause := fmt.Sprintf(`f."%s" %s ?%s`, columnName, opString, escapeString)
+		return clause, []any{formatMatchTarget(filter)}, nil
+	case NotEq:
+		if filter.Partial {
+			opString = "NOT LIKE"
+			escapeString = escapeBackslashDirective
+		} else {
+			opString = "!="
+		}
+		clause := fmt.Sprintf(`f."%s" %s ?%s`, columnName, opString, escapeString)
+		return clause, []any{formatMatchTarget(filter)}, nil
+	case Exists:
+		clause := fmt.Sprintf(`f."%s" IS NOT NULL`, columnName)
+		return clause, []any{}, nil
+	case NotExists:
+		clause := fmt.Sprintf(`f."%s" IS NULL`, columnName)
+		return clause, []any{}, nil
+	case In:
+		fallthrough
+	case NotIn:
+		target := "()"
+		if len(filter.Matches) > 0 {
+			target = fmt.Sprintf("(?%s)", strings.Repeat(", ?", len(filter.Matches)-1))
+		}
+		opString = "IN"
+		if filter.Op == NotIn {
+			opString = "NOT IN"
+		}
+		clause := fmt.Sprintf(`f."%s" %s %s`, columnName, opString, target)
+		matches := make([]any, len(filter.Matches))
+		for i, match := range filter.Matches {
+			matches[i] = match
+		}
+		return clause, matches, nil
+	}
+
+	return "", nil, fmt.Errorf("unrecognized operator: %s", opString)
+}
+
+func (l *ListOptionIndexer) getLabelFilter(filter Filter, dbName string) (string, []any, error) {
+	opString := ""
+	escapeString := ""
+	if len(filter.Field) < 3 || filter.Field[0] != "metadata" || filter.Field[1] != "labels" {
+		return "", nil, fmt.Errorf("expecting a metadata.labels field, got '%s'", strings.Join(filter.Field, "."))
+	}
+	matchFmtToUse := strictMatchFmt
+	labelName := filter.Field[2]
+	switch filter.Op {
+	case Eq:
+		if filter.Partial {
+			opString = "LIKE"
+			escapeString = escapeBackslashDirective
```

Comment:
How about inlining the directive in the SQL query where it's used, instead of having this as a constant copied into a variable and then into the query?

(yes, I understand that's a bit of repetition, maybe it's just me suffering from indirection motion sickness!)

PS. same about matchFmt. I had a hard time looking into 4 places before coalescing the sense of the query in my mind while reviewing 🦀

Reply:
# TBD
We need to hear from Richard if this business with quoting strings to be exact vs doing substring matches is going to survive.

Reply:
ACK

Reply:
There needs to be a way to differentiate between partial and exact matches. As long as that's there am open to offers on syntax (UI changes would be straight forward)

Reply:
@richard-cox please note that the context for this thread is labels.

My understanding is that, for labels, you need power to match [label selectors](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/).

Now label selectors do implement equality, which we call exact for normal fields. They do not implement "substring equality".

Can you confirm we will continue to need both "exact" and "substring" equality for normal fields, and label selector power on label fields, which _does not include "substring" equality_?

Reply:
```
Can you confirm we will continue to need both "exact" and "substring" equality for normal fields, and label selector power on label fields, which does not include "substring" equality?
```

@moio correct.

Label selectors are exact/not exact or in/not in a set. Filtering outside of labels would need to cover exact/not exact as well as substrings ... but not sets

Reply:
@ericpromislow please keep in mind the UI needs to evaluate results of a selector from a Kubernetes resource "as it is" (eg. taking a .spec.selector out of a Deployment, pass it to Steve to get a list of targeted pods in order to display a preview - If the user interactively changes the selector, update the preview on-the-fly).

Therefore there should be a way to pass in a plain Kubernetes selector as a filter with the exact same syntax and get the exact same semantics in return.

It is my understanding from Steve code that we are now parsing `filter=` query parameters with a syntax similar to vanilla Kubernetes selectors, with an extension for quoted strings (which are interpreted as exact matches, while the default is substring matching). Note that is different from standard Kubernetes selectors, where matching is exact by default (and substring matching does not even exist).

If the above is right (please correct me otherwise!) you might want to consider adding a hard distinction between `filter=` parameters - to be kept working exactly as today, and a new kind of parameters - say, `selectorFilter=` - that recognizes and implements Kubernetes selector syntax and semantics 1:1.

Please let me know what you think

Reply:
It is correct that if the target of a `=` or `!=` operator is quoted, we do an exact match. Otherwise we do a substring match. This is an extension of the standard kubectl label selector parser, which is why I forked and modified it.

I'd rather not make any more changes until Richard is able to integrate these changes and see if they work as expected from the UI. I don't know if the UI in 2.10 releases is able to issue these kinds of filters -- I haven't seen any evidence of it.

Reply:
From a UI perspective splitting these into two non-overlapping filters would work. One covering the existing `filter` with our own syntax and another covering the kube label select (and standard kube filtering?) api syntax. It feels like the latter is a more neater solution, though i haven't looked through the changes in this PR.

The UI effort to cover either approach (single uber filter or two filters) is small.

Eric you're right, there's currently no way to exercise the label selector via api in the ui. Given an image with your changes in (that supports amd/x86) i could implement the feature in a single area and knock up a build that can very easily be used (by changing two rancher management.cattle.io `setting` values)